### PR TITLE
Minimal implementation of MP when using SP.

### PR DESCRIPTION
**Pull request by SparkiDev**
--enable-sp-math to include minimal implementation of MP (only with
--enable-sp.)
Add futher functionality for ECC (conditionally compiled):
- check key
- is point on curve
- API to add and double projective points
- API to map from project to affine
- Uncompress point (including sqrt)
Some configuration options will not work with SP math - configure.ac
detects this and errors out.
Change test code to better support SP sizes only.

### CODE REVIEW

File: wolfcrypt/src/sp_int.c [hunk 1]
```diff
@@ -0,0 +1,486 @@
...
+int sp_leading_bit(sp_int* a)
+{
+    int bit = 0;
+    sp_int_digit d;
+
+    if (a->used > 0) {
+        d = a->dp[a->used - 1];
+        while (d > (sp_int_digit)0xff)
+            d >>= 8;
+        bit = d >> 7;
```
Comment (dgarske) [inline wolfcrypt/src/sp_int.c]:
```
wolfcrypt/src/sp_int.c:279:17: error: implicit conversion loses integer precision: 'sp_int_digit' (aka 'unsigned long long') to 'int' [-Werror,-Wshorten-64-to-32]
        bit = d >> 7;
            ~ ~~^~~~
```

File: wolfcrypt/src/sp_int.c [hunk 2]
```diff
@@ -0,0 +1,486 @@
...
+    for (; i < a->used; i++) {
+        r->dp[i] = a->dp[i] + c;
+        c = r->dp[i] == 0;
+    }
+    for (; i < b->used; i++) {
+        r->dp[i] = b->dp[i] + c;
+        c = r->dp[i] == 0;
+    }
+    r->dp[i] = c;
+    a->used = i + c;
```
Comment (dgarske) [inline wolfcrypt/src/sp_int.c]:
```
wolfcrypt/src/sp_int.c:468:17: error: implicit conversion loses integer precision: 'long long' to 'int' [-Werror,-Wshorten-64-to-32]
    a->used = i + c;
            ~ ~~^~~
```

**Comment by dgarske** [comment]
Sean, For sp_int.h we need to inplement the WOLFSSL_PUBLIC_MP logic I mentioned in the review. Then this will be ready for primetime.

File: wolfssl/wolfcrypt/sp_int.h [hunk 3]
```diff
@@ -90,27 +95,27 @@ typedef struct sp_int {
...
-WOLFSSL_LOCAL int sp_set(sp_int* a, sp_int_digit d);
-WOLFSSL_LOCAL int sp_iszero(sp_int* a);
-WOLFSSL_LOCAL void sp_clamp(sp_int* a);
-WOLFSSL_LOCAL int sp_grow(sp_int* a, int l);
-WOLFSSL_LOCAL void sp_zero(sp_int* a);
-WOLFSSL_LOCAL int sp_add_d(sp_int* a, sp_int_digit d, sp_int* r);
-WOLFSSL_LOCAL int sp_lshd(sp_int* a, int s);
-WOLFSSL_LOCAL int sp_add(sp_int* a, sp_int* b, sp_int* r);
+MP_API int sp_init(sp_int* a);
+WOLFSSL_API int sp_init_multi(sp_int* a, sp_int* b, sp_int* c, sp_int* d,
```
Comment (dgarske) [inline wolfssl/wolfcrypt/sp_int.h]: Sean this should MP_API now. The API unit test only requires mp_init_multi when WOLFSSL_PUBLIC_MP is defined. Thanks

Comment (SparkiDev) [reply wolfssl/wolfcrypt/sp_int.h]: Fixed

**Pull request merged**